### PR TITLE
feat(images): adds a new image manipulation service

### PR DESCRIPTION
**Pull request by hypeJunction**
Adds a basic image manipulation service based on Imagine (temporarily supporting only GD interface).
Entity icons are now cropped via the new image service

### CODE REVIEW

File: engine/classes/Elgg/Di/ServiceProvider.php [hunk 1]
```diff
@@ -232,6 +233,11 @@ public function __construct(\Elgg\Config $config) {
 
 		$this->setClassName('input', \Elgg\Http\Input::class);
 
+		$this->setFactory('imageService', function(ServiceProvider $c) {
+			$imagine = new \Imagine\Gd\Imagine();
```
Comment (hypeJunction) [inline engine/classes/Elgg/Di/ServiceProvider.php]: We can later a config option to use Imagick or Gdmagick

File: engine/classes/Elgg/EntityIconService.php [hunk 2]
```diff
@@ -263,22 +260,24 @@ public function saveIcon(ElggEntity $entity, ElggFile $file, $type = 'icon', arr
...
-				$icon->close();
-				unset($image_bytes);
-			} else {
-				$this->logger->error("Failed to create {$size} icon from {$file->getFilenameOnFilestore()} with coords [{$x1}, {$y1}],[{$x2}, {$y2}]");
+			$source = $file->getFilenameOnFilestore();
+			$destination = $icon->getFilenameOnFilestore();
+
+			$resize_params = array_merge($opts, $coords);
+			
+			if (!_elgg_services()->imageService->resize($source, $destination, $resize_params)) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/EntityIconService.php]: I am thinking we can add a batch method, so we don't have to open/close the file handle for each icon

File: engine/classes/Elgg/ImageService.php [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+		$square = elgg_extract('square', $params, false);
+		$upscale = elgg_extract('upscale', $params, false);
+
+		$x1 = (int) elgg_extract('x1', $params, 0);
+		$y1 = (int) elgg_extract('y1', $params, 0);
+		$x2 = (int) elgg_extract('x2', $params, 0);
+		$y2 = (int) elgg_extract('y2', $params, 0);
+
+		$crop_width = $x2 - $x1;
```
Comment (hypeJunction) [inline engine/classes/Elgg/ImageService.php]: Remove

**Comment by mrclay** [comment]
Do we need to update all these deps? Would rather make that decision in a separate PR.

**Comment by hypeJunction** [comment]
I don't think we can actually add a dependency without updating all others...

**Comment by hypeJunction** [comment]
I guess require should work. Trying now.

**Comment by hypeJunction** [comment]
Nope, that still updates all deps:

```sh
$ composer require imagine/imagine
You are running composer with xdebug enabled. This has a major impact on runtime performance. See https://getcomposer.org/xdebug
Using version ^0.6.3 for imagine/imagine
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Removing composer/installers (v1.2.0)
  - Installing composer/installers (v1.1.0)
    Loading from cache

  - Removing doctrine/dbal (v2.5.5)
  - Installing doctrine/dbal (v2.5.4)
    Loading from cache

  - Removing symfony/http-foundation (v2.8.11)
  - Installing symfony/http-foundation (v2.8.9)
    Loading from cache

  - Removing bower-asset/requirejs (2.3.2)
  - Installing bower-asset/requirejs (2.2.0)
    Loading from cache

  - Removing sebastian/environment (1.3.8)
  - Installing sebastian/environment (1.3.7)
    Loading from cache

  - Removing symfony/yaml (v3.1.4)
  - Installing symfony/yaml (v3.1.3)
    Loading from cache

Writing lock file
Generating autoload files
```

**Comment by mrclay** [comment]
That's after you reset (don't delete) the lock?

**Comment by mrclay** [comment]
revert the lock to 2.x version, then `composer require imagine/imagine ^0.6.3`

**Comment by hypeJunction** [comment]
I am not deleting it. It's in 2.x base.

**Comment by hypeJunction** [comment]
Same thing.

**Comment by hypeJunction** [comment]
I think it worked.

**Comment by hypeJunction** [comment]
I think this is ready to go in, if anyone wants to review.

File: docs/guides/upgrading.rst [hunk 4]
```diff
@@ -50,6 +53,13 @@ New API for working with file uploads
  * ``elgg_get_uploaded_files()`` - returns an array of Symfony uploaded file objects
  * ``ElggFile::acceptUploadedFile()`` - moves an uploaded file to Elgg's filestore
 
+New API for manipulating images
+-------------------------------
+
+New image manipulation service implements a more efficient approach to cropping and resizing images.
```
Comment (mrclay) [inline docs/guides/upgrading.rst]: By "more efficient" do you mean simpler API or do we know it's faster, too?

**Review by mrclay — APPROVED** [review]

File: engine/classes/Elgg/EntityIconService.php [hunk 5]
```diff
@@ -263,22 +260,24 @@ public function saveIcon(ElggEntity $entity, ElggFile $file, $type = 'icon', arr
 			}
 
 			$icon = $this->getIcon($entity, $size, $type);
-
+			$icon->open('write');
+			$icon->close();
```
Comment (jdalsem) [inline engine/classes/Elgg/EntityIconService.php]: Why is this open/close needed? Maybe add a comment for it?

File: engine/lib/filestore.php [hunk 6]
```diff
@@ -54,6 +54,34 @@ function get_uploaded_file($input_name) {
...
+ *                               coordinates. The source image will first be cropped
+ *                               to these coordinates, and then resized to match
+ *                               width/height parameters
+ *                             - BOOL 'square' - square images will fill the
+ *                               bounding box (width x height). In Imagine's terms,
+ *                               this equates to OUTBOUND mode
+ *                             - BOOL 'upscale' - if enabled, smaller images
+ *                               will be upscaled to fit the bounding box.
+ * @return bool
+ */
```
Comment (jdalsem) [inline engine/lib/filestore.php]: add `@since`

**Review by jdalsem — COMMENTED** [review]
just the

**Comment by jdalsem** [comment]
don't know why i typed 'just the'... probably wanted to say 'just the `@since`... the rest LGTM'

File: docs/guides/upgrading.rst (hunk 4 above)
Comment (hypeJunction) [inline docs/guides/upgrading.rst]: It's uses less memory as it doesn't have to keep all the temporary data/raw bytes in memory. Plus we can switch to imagick.

**Comment by hypeJunction** [comment]
@jdalsem Addressed your comments. Pull the plug if you are happy.

**Review by jdalsem — APPROVED** [review]

**Pull request merged**